src/util/isDomainSpecifiedByUser.ts

```diff
@@ -28,3 +30,198 @@ export function isDomainSpecifiedByUser(
+      } catch (ex) {
+        /* ignore the exception and compute domain from data later */
+      }
+    }
+
+    const candidate = [finalMin, finalMax];
+    if (isWellFormedNumberDomain(candidate)) {
+      return candidate;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * So Recharts allows users to provide their own domains,
+ * but it also places some expectations on what the domain is.
+ * We can improve on the typescript typing, but we also need a runtime test
+ * to observe that the user-provided domain is well-formed,
+ * that is: an array with exactly two numbers.
+ * If the user-provided domain is not well-formed,
+ * this function will return undefined - in which case we should traverse the data to calculate the real domain.
+ *
+ * This function is for parsing the numerical domain only.
+ *
+ * @param userDomain external prop, user provided, before validation. Can have various shapes: array, function, special magical strings inside too.
+ * @param dataDomain calculated from data. Can be undefined, as an option for performance optimization
+ * @param allowDataOverflow provided by users. If true then the data domain wins
+ * @param allowDecimals if true then recharts is allowed to generate domain with decimal numbers; if false the numbers are always integers
+ *
+ * @return [min, max] domain if it's well-formed; undefined if the domain is invalid
+ */
+export function parseNumericalUserDomain(
+  userDomain: AxisDomain | undefined,
+  dataDomain: NumberDomain | undefined,
+  allowDataOverflow: boolean,
+  allowDecimals: boolean,
+): NumberDomain | undefined {
+  if (!allowDataOverflow && dataDomain == null) {
+    // Cannot compute data overflow if the data is not provided
+    return undefined;
+  }
+  if (typeof userDomain === 'function') {
+    try {
+      const result = userDomain(dataDomain, allowDataOverflow);
+      if (isWellFormedNumberDomain(result)) {
+        return extendDomain(result, dataDomain, allowDataOverflow);
+      }
+    } catch (ex) {
+      /* ignore the exception and compute domain from data later */
+    }
+  }
+  if (Array.isArray(userDomain) && userDomain.length === 2) {
+    const [providedMin, providedMax] = userDomain;
+    let finalMin, finalMax: number, niceTicks: ReadonlyArray<number>;
+
+    if (providedMin === 'auto') {
+      if (dataDomain != null) {
+        niceTicks = getNiceTickValues(dataDomain, undefined, allowDecimals);
+        finalMin = Math.min(...niceTicks);
+      }
+    } else if (typeof providedMin === 'number' && !Number.isNaN(providedMin)) {
+      finalMin = providedMin;
+    } else if (typeof providedMin === 'function') {
+      try {
+        finalMin = providedMin(dataDomain?.[0]);
+      } catch (ex) {
+        /* ignore the exception and compute domain from data later */
+      }
+    } else if (typeof providedMin === 'string' && MIN_VALUE_REG.test(providedMin)) {
+      const value = +MIN_VALUE_REG.exec(providedMin)[1];
+      finalMin = dataDomain?.[0] - value;
```

Comment:
Doesn't the strings that are allowed allow for math? Like more than subtraction you could do "dataMin + 400" and get that as your data min. Or "dataMax - 400"

Your test cases seem to indicate that doesn't work. Maybe we should invariant somewhere in here if people do something wrong like that? I guess we're not trying to change current behavior here..

Reply:
If you look closely on the regular expressions you will find that they allow expanding the domain but not contracting: https://github.com/recharts/recharts/blob/6e9f9fe5b06ab939135a48f4c68e124d89e3f337/src/util/ChartUtils.ts#L1295

I was a bit surprised by that but I decided against changing the regex, and I copy pasted this chunk of code from its original location: https://github.com/recharts/recharts/blob/6e9f9fe5b06ab939135a48f4c68e124d89e3f337/src/util/ChartUtils.ts#L1312

Reply:
I have some test cases for math that expands the domain: https://github.com/PavelVanecek/recharts/blob/9c7a275b59887ac2c649228ef526169d22bade16/test/util/isDomainSpecifiedByUser.spec.ts#L140

And also for math that contracts the domain: https://github.com/PavelVanecek/recharts/blob/9c7a275b59887ac2c649228ef526169d22bade16/test/util/isDomainSpecifiedByUser.spec.ts#L132

Reply:
Cool cool, I figured you didn't change it I just always tried to use it this way myself and totally thought it worked 😂. All good as long as the function versions allow you to do whatever you want

Reply:
With the test suite it should now be much easier to add new behaviour. I'm expecting that one of the 500 open issues is asking for that anyway.

But. Let's release 3.0 first!